libgit/autogit_node_wrappers.go

```diff
@@ -28,21 +31,217 @@ import (
 //   `readonlyNode`, and an `tlfTypeNode`.
 // * `tlfTypeNode` allows the auto-creation of subdirectories
 //   representing TLFs, e.g. .kbfs_autogit/private/max or
-//   .kbfs_autogit/team/keybase.  It wraps child nodes as a
-//   `readOnlyNode`.  TODO(KBFS-2678): allow repo autocreation under
-//   a `tlfTypeNode`.
+//   .kbfs_autogit/team/keybase.  It wraps child nodes in two ways, as
+//   a `readOnlyNode` and a `tlfNode`.
+// * `tlfNode` allows auto-creation of subdirectories representing
+//   valid repository checkouts of the corresponding TLF, e.g.
+//   `.kbfs_autogit/private/chris/dotfiles`.  It wraps child nodes in
+//   two ways, as both a `readonlyNode` and a `repoNode`.
+// * `repoNode` allow auto-clone and auto-pull of the corresponding
+//   repository on its first access.  When the directory corresponding
+//   to the node is read for the first time for this KBFS instance,
+//   the `repoNode` asks `AutogitManager` to either kick off a clone
+//   or a pull for the repository in question, which will checkout a
+//   copy of the source repo under that directory.  The operation will
+//   block on that clone/pull operation until it is finished, until
+//   the given context is canceled, or until 10 seconds is up.  But if
+//   it doesn't finish in time, the operation continues in the
+//   background and should update the directory asynchronously.  If
+//   the operation is a clone, a "CLONING" file will be visible in the
+//   directory until the clone completes.  `repoNode` wraps each child
+//   node as a `readonlyNode`.
 
 type ctxReadWriteKeyType int
+type ctxSkipPopulateKeyType int
 
 const (
-	autogitRoot                         = ".kbfs_autogit"
-	ctxReadWriteKey ctxReadWriteKeyType = 1
+	autogitRoot = ".kbfs_autogit"
+
+	populateTimeout = 10 * time.Second
+
+	ctxReadWriteKey    ctxReadWriteKeyType    = 1
+	ctxSkipPopulateKey ctxSkipPopulateKeyType = 1
 
 	public  = "public"
 	private = "private"
 	team    = "team"
 )
 
+type repoNode struct {
+	libkbfs.Node
+	am       *AutogitManager
+	h        *libkbfs.TlfHandle
+	repoName string
+
+	lock                 sync.Mutex
+	populated            bool
+	populatingInProgress chan struct{}
+}
+
+var _ libkbfs.Node = (*repoNode)(nil)
+
+func newRepoNode(n libkbfs.Node, am *AutogitManager, h *libkbfs.TlfHandle,
+	repoName string) *repoNode {
+	return &repoNode{
+		Node:     n,
+		am:       am,
+		h:        h,
+		repoName: repoName,
+	}
+}
+
+func (rn *repoNode) dstDir() string {
+	var typeStr string
+	switch rn.h.Type() {
+	case tlf.Public:
+		typeStr = public
+	case tlf.Private:
+		typeStr = private
+	case tlf.SingleTeam:
+		typeStr = team
+	}
```

Comment:
There is a switch like that in the `tlf` package, but it names `tlf.SingleTeam` to be "singleTeam" instead of "team", to match the type for debugging purposes.  (Someday we might have multiteam folders, that could still show up under "team" subdirectory.)

So far we've been working on the idea that other packages are responsible for their own presentation layer -- i.e., libfs, simplefs, and libgit are all allowed to show/accept the different tlf types as different strings, and that's ok.  We could consolidate that I suppose, but that hasn't been that plan so far.

Reply:
I understand, but it seems like the `/keybase/{private,public,team}` designation is pretty standardized, and should be owned by the `tlf` package.

Reply:
I don't really see that as obvious, actually.  The current way still makes more sense to me.   Maybe file another ticket for more discussion, I'm not in inclined to put it in this PR.